Neste guided project, trabalharemos com um conjunto de dados de carros usados do eBay kleinanzeigen, uma seção de classificados do site alemão eBay.
O objetivo deste projeto é limpar os dados e analisar as  listagens de carros usados incluídas.

In [23]:
#Importando as bibliotecas necessárias.
import numpy as np
import pandas as pd

#Lendo o dataset autos.csv e armazenando seus dados na váriavel autos.
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [24]:
#Impriminto das 5 primeiras linha do dataset e as principais informações.
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [25]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

Observacoes:
    1 - O conjunto de dados contém 20 colunas e a maioria das delas sao strings.
    2 - Os nomes das colunas usam camelcase em vez do snakecase preferido do Python, portanto, significa que nao podemos simplesmente substituir espacos por sublinhados.  
    3 - Os objetos da coluna name, apresentam muitos caracteres especiais e variacoes entre maiusculas e minusculas.
    4 - Os objetos da coluna price podem ser convertidos para float e a unidade dos valores expecificados no nome da coluna.
    5 - Os objetos da coluna odometer podem ser convertidos para float e a unidade dos valores expecificados no nome da coluna.

In [26]:
#Vamos converter os nomes das colunas do padrão camelcase para o padrão snakecase.
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [27]:
map = {
    'yearOfRegistration':'registration_year',
    'monthOfRegistration':'registration_month',
    'notRepairedDamage':'unrepaired_damage',
    'dateCreated':'ad_created',
    'dateCrawled':'date_crawled',
    'offerType':'offer_type',
    'vehicleType':'vehicle_type',
    'powerPS':'power_ps',
    'fuelType':'fuel_type',
    'nrOfPictures':'nr_pictures',
    'postalCode':'postal_code',
    'lastSeen':'last_seen'
}
autos.rename(columns=map, inplace=True)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Agora vamos fazer uma exploração dos dados básicos para determinar quais outras tarefas de limpeza precisam ser feitas.

In [28]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

Após a análise dos dados, notou-se que as colunas nr_pictures, offer_type e seller apresentam muitos valores correspondentes e nulos portanto, para preservar nossas analises futuras, elas devem ser removidas. Além disso, price e odometer são valores numéricos armazenados como texto.

In [29]:
#Removendo as colunas com valores semelhantes e nulos.
autos = autos.drop(columns=['nr_pictures','offer_type','seller'])

#Remoção dos caracteres extras e conversão das colunas númericas armazenadas em texto.
autos['price'] = (autos['price'].str.replace('$','').str.replace(',', '').astype(float))
autos['odometer'] = (autos['odometer'].str.replace('km','').str.replace(',', '').astype(float))

#No caso da coluna odometer nao podemos perder a informacao que os valores sao medidos em km, vamos adicionar a informacao ao novo da coluna.
autos.rename(columns={"odometer":'odometer_km'},inplace=True)

In [30]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0  5000.0  control          bus               2004    manuell       158   
1  8500.0  control    limousine               1997  automatik       286   
2  8990.0     test    limousine               2009    manuell       102   
3  4350.0  control   kleinwagen               2007  automatik        71   
4  1350.0     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere     150000.0                   3       lpg     peugeot   
1     7er     150000.0                   6    benzin         bmw   
2    golf      70000.0                   7    benzin  volkswagen   
3  fortwo      70000.0                   6    benzin       smart   
4   focus     150000.0                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

Vamos continuar explorando os dados, especificamente procurando dados que não parecem corretos. Iniciando com as colunas odometer_km.

In [31]:
autos.odometer_km.unique().shape

(13,)

In [32]:
autos.odometer_km.describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [33]:
autos.odometer_km.value_counts().sort_index(ascending=True)

5000.0        967
10000.0       264
20000.0       784
30000.0       789
40000.0       819
50000.0      1027
60000.0      1164
70000.0      1230
80000.0      1436
90000.0      1757
100000.0     2169
125000.0     5170
150000.0    32424
Name: odometer_km, dtype: int64

A coluna odometer não apresenta valores irrealistas.
Agora vamos análisar a coluna price.

In [34]:
autos.price.unique().shape

(2357,)

In [35]:
autos.price.describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [36]:
autos.price.value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

Após a análise dos valores de forma ascendente, percebemos uma inconsistência a partir do valor 350000.0 

In [37]:
autos.price.value_counts().sort_index(ascending=True).head(20)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price, dtype: int64

Já na analise descendente, percebemos uma grande quantidade de 0, portanto agora vamos remover os valores maiores que 350000.0 e menores que 1.

In [38]:
#Eliminando os outliers
autos = autos[autos.price.between(1,350000)]

Verificando os dados após a remoção.

In [39]:
autos.price.value_counts().sort_index(ascending=True).head()

1.0    156
2.0      3
3.0      1
5.0      2
8.0      1
Name: price, dtype: int64

In [40]:
autos.price.value_counts().sort_index(ascending=False).head()

350000.0    1
345000.0    1
299000.0    1
295000.0    1
265000.0    1
Name: price, dtype: int64

Vamos agora passar  para as colunas de datas e entender o intervalo de dados abordados.

In [41]:
autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [42]:
autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

In [43]:
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

Após a análise das distribuições em porcentagem dos 10 primeiros caracteres das colunas date_crawled, ad_created e last_seen, não foram notadas irregularidades que necessitarem de correções.

In [44]:
autos.registration_year.describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

A coluna registration_year apresenta incoerências. Carros registrados nos anos 1000 e 9999 são considerados outliers, precisam ser tratados.

In [47]:
autos.registration_year.value_counts().sort_index(ascending=True).head(10)

1000    1
1001    1
1111    1
1800    2
1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
Name: registration_year, dtype: int64

In [49]:
autos.registration_year.value_counts().sort_index(ascending=False).head(15)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
Name: registration_year, dtype: int64

Como um carro não pode ser registrado depois que a listagem foi vista, qualquer veículo com ano de registro maior que 2016 deve ser removido. Já determinar o primeiro ano válido é mais complicado, vamos tomar como partida o ano de 1900.  

In [52]:
#Eliminando os outliers e apresentando a distribuição dos valores restantes.
autos = autos[autos.registration_year.between(1910,2016)]
autos.registration_year.value_counts(normalize=True)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
          ...   
1938    0.000021
1948    0.000021
1927    0.000021
1931    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

Ao trabalharmos com dados sobre carros, é importante análisarmos variações entre diferentes marcas. Podemos usar a técnica de agregação entender a coluna  brand.

In [55]:
autos['brand'].value_counts().head(20)

volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
Name: brand, dtype: int64

In [58]:
#Criando uma serie com as 20 marcas mais populares da lista.
top_brand=autos.brand.value_counts().head(20).index
top_brand

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')

In [61]:
#Criando os dicionários para os preços médios e as quilometragens médias das top 20 marcas.
dict_mean_price={}
dict_mean_odometer={}
#Interando sobre eles.
for brand in top_brand:
    mean_price=autos[autos['brand']==brand]['price'].mean()
    mean_odometer=autos[autos['brand']==brand]['odometer_km'].mean()
    dict_mean_price[brand]=mean_price
    dict_mean_odometer[brand]=mean_odometer

In [62]:
#Convertendo os dicionários em séries de objetos.
series_mean_price=pd.Series(dict_mean_price)
series_mean_odometer=pd.Series(dict_mean_odometer)

In [63]:
#Criando um dataframe com a serie de objetos de preços médios.
df_relation_brand=pd.DataFrame(series_mean_price,columns=['mean_price'])
df_relation_brand

mean_price
volkswagen       5402.410262
bmw              8332.820518
opel             2975.241935
mercedes_benz    8628.450366
audi             9336.687454
ford             3749.469507
renault          2474.864607
peugeot          3094.017229
fiat             2813.748538
seat             4397.230950
skoda            6368.000000
nissan           4743.402525
mazda            4112.596615
smart            3580.223903
citroen          3779.139144
toyota           5167.091062
hyundai          5365.254274
sonstige_autos  12338.550218
volvo            4946.501171
mini            10613.459658

In [66]:
#Atribuindo a serie de objetos de kilometragens médios, como uma nova coluna.
df_relation_brand['mean_odometer_km']=series_mean_odometer
df_relation_brand.sort_index()

mean_price  mean_odometer_km
audi             9336.687454     129157.386785
bmw              8332.820518     132572.513140
citroen          3779.139144     119694.189602
fiat             2813.748538     117121.971596
ford             3749.469507     124266.012872
hyundai          5365.254274     106442.307692
mazda            4112.596615     124464.033850
mercedes_benz    8628.450366     130788.363313
mini            10613.459658      88105.134474
nissan           4743.402525     118330.995792
opel             2975.241935     129310.035842
peugeot          3094.017229     127153.625269
renault          2474.864607     128071.331213
seat             4397.230950     121131.301290
skoda            6368.000000     110848.563969
smart            3580.223903      99326.777610
sonstige_autos  12338.550218      89956.331878
toyota           5167.091062     115944.350759
volkswagen       5402.410262     128707.158791
volvo            4946.501171     138067.915691

Após a análise dos preços médios e quilometragens méidas das 20 marcas mais comuns,  não podemos perceber uma relação direta entre os dois valores.